include/RAJA/pattern/params/reducer.hpp

```diff
@@ -101,48 +95,94 @@ namespace detail
-    RAJA_HOST_DEVICE ARG_TUP_T get_lambda_arg_tup() { return camp::make_tuple(&val); }
+    // These are types and parameters extracted from this struct, and given to the forall.
+    using ARG_TUP_T = camp::tuple<VType*>;
+    RAJA_HOST_DEVICE ARG_TUP_T get_lambda_arg_tup() { return camp::make_tuple(&valop_m); }
 
     using ARG_LIST_T = typename ARG_TUP_T::TList;
     static constexpr size_t num_lambda_args = camp::tuple_size<ARG_TUP_T>::value ;
   };
 
-} // namespace detail
+  // Partial specialization of Reducer for ValLoc
+  // T must be a basic data type
+  // VType must be ValOp<ValLoc<T,IndexType>,Op>
+  template <template <typename, typename, typename> class Op, typename T, typename IndexType>
+  struct Reducer<Op, T, IndexType, ValOp<ValLoc<T,IndexType>, Op>> : public ForallParamBase {
+    using value_type = ValLoc<T,IndexType>; // Note that this struct operates on ValLoc
+    using op = Op<value_type,value_type,value_type>;
+    using VType = ValOp<ValLoc<T,IndexType>, Op>;
 
-template <template <typename, typename, typename> class Op, typename T>
-auto constexpr Reduce(T *target)
-{
-  return detail::Reducer<Op<T, T, T>, T>(target);
-}
+    Reducer() = default;
 
+    // ValLoc constructor
+    // Note that the passthru variables point to the val and loc within the user defined target ValLoc
+    RAJA_HOST_DEVICE Reducer(value_type *target_in) : valop_m(VType{}), target(target_in), passthruval(&target_in->val), passthruindex(&target_in->loc) {}
 
+    // Pass through constructor for ReduceLoc<>(data, index) case
+    // The passthru variables point to vars defined by the user
+    RAJA_HOST_DEVICE Reducer(T *data_in, IndexType *index_in) : valop_m(VType(*data_in, *index_in)), target(&valop_m.val), passthruval(data_in), passthruindex(index_in) {}
 
-namespace detail
-{
+    Reducer(Reducer const &) = default;
+    Reducer(Reducer &&) = default;
+    Reducer& operator=(Reducer const &) = default;
+    Reducer& operator=(Reducer &&) = default;
 
-  //
-  //
-  // Basic ReducerLoc
-  //
-  //
-  template <typename Op, typename T>
-  struct ReducerLoc : public Reducer<Op, T> {
-    using Base = Reducer<Op, T>;
-    using value_type = typename Base::value_type;
-    ReducerLoc(value_type *target_in) {
-      Base::target = target_in;
-      Base::val = value_type(Op::identity());
-    }
+    VType valop_m = VType{};
+
+    value_type *target = nullptr;
+
+    T *passthruval = nullptr;
+    IndexType *passthruindex = nullptr;
+
+    // This single-argument set() matches the basic data type Reducer::set()'s
+    // function signature to make the init/combine/resolve backend calls easier
+    // to maintain, but operates on the passthru variables instead
+    RAJA_HOST_DEVICE void set(value_type in) { *passthruval = in.val; *passthruindex = in.loc; }
+
+    RAJA_HOST_DEVICE
+    value_type &
+    getVal() { return valop_m.val; }
+
+#if defined(RAJA_CUDA_ACTIVE) || defined(RAJA_HIP_ACTIVE) || defined(RAJA_SYCL_ACTIVE)
```

Comment:
I think we have a macro that checks all these.

Reply:
Doesn't look we like we do. This only happens in this file, so I'll leave it in.